src/tools/miri/src/bin/miri.rs

```diff
@@ -216,77 +219,41 @@ fn init_late_loggers(tcx: TyCtxt<'_>) {
-                    )
-                }
-            }
-            format!("{}/toolchains/{}", home, toolchain)
-        }
-        _ => option_env!("RUST_SYSROOT")
-            .unwrap_or_else(|| {
-                show_error!(
-                    "To build Miri without rustup, set the `RUST_SYSROOT` env var at build time",
-                )
-            })
-            .to_owned(),
-    })
-}
-
 /// Execute a compiler with the given CLI arguments and callbacks.
 fn run_compiler(
     mut args: Vec<String>,
     target_crate: bool,
     callbacks: &mut (dyn rustc_driver::Callbacks + Send),
 ) -> ! {
-    // Make sure we use the right default sysroot. The default sysroot is wrong,
-    // because `get_or_default_sysroot` in `librustc_session` bases that on `current_exe`.
-    //
-    // Make sure we always call `host_sysroot` as that also does some sanity-checks
-    // of the environment we were built in and whether it matches what we are running in.
-    let host_default_sysroot = host_sysroot();
-    // Now see if we even need to set something.
     let sysroot_flag = "--sysroot";
     if !args.iter().any(|e| e == sysroot_flag) {
         // No sysroot was set, let's see if we have a custom default we want to configure.
         let default_sysroot = if target_crate {
             // Using the built-in default here would be plain wrong, so we *require*
             // the env var to make sure things make sense.
-            Some(env::var("MIRI_SYSROOT").unwrap_or_else(|_| {
+            env::var("MIRI_SYSROOT").unwrap_or_else(|_| {
                 show_error!(
                     "Miri was invoked in 'target' mode without `MIRI_SYSROOT` or `--sysroot` being set"
                 )
-            }))
+            })
         } else {
-            host_default_sysroot
+            let sysroot_compile_time_args = SysrootCompileTimeArgsForHost {
+                sysroot: option_env!("RUST_SYSROOT").map(ToString::to_string),
+                rustup_home: option_env!("RUSTUP_HOME").map(ToString::to_string),
+                multirust_home: option_env!("MULTIRUST_HOME").map(ToString::to_string),
+                rustup_toolchain: option_env!("RUSTUP_TOOLCHAIN").map(ToString::to_string),
+                multirust_toolchain: option_env!("MULTIRUST_TOOLCHAIN").map(ToString::to_string),
+            };
+
+            rustc_session::filesearch::get_or_default_sysroot(Some(sysroot_compile_time_args))
+                .map(|pb| pb.to_string_lossy().to_string())
+                .expect(
+                    "To build Miri without rustup, set the `RUST_SYSROOT` env var at build time",
+                )
         };
-        if let Some(sysroot) = default_sysroot {
-            // We need to overwrite the default that librustc_session would compute.
-            args.push(sysroot_flag.to_owned());
-            args.push(sysroot);
-        }
+
+        args.push(sysroot_flag.to_owned());
+        args.push(default_sysroot);
```

Comment:
This is still using the `--sysroot` arg-passing hack to set the sysroot, so I don't see how that's better than the status quo.

All this seems to do is extract some common code from Miri and clippy, but we could have that more easily by moving that code to a helper crate. The actual problem of rustc hard-coding the default sysroot does not get resolved by this.

Reply:
> This is still using the --sysroot arg-passing hack to set the sysroot, so I don't see how that's better than the status quo.

I suppose I misunderstand the goal here then. I guess all sysroot logic has to be excluded from miri and it should still keep working as before. If so, I will work on that.